File: lib/roundtable/output.ex
defmodule Roundtable.Output do
  @moduledoc "Builds normalized result maps and encodes final JSON output."

  @spec build_result(
          cli_name :: String.t(),
          path :: String.t() | nil,
          model :: String.t() | nil,
          probe_result :: map() | nil,
          run_result :: map() | nil,
          cli_module :: module()
        ) :: map()
  def build_result(cli_name, nil, model, _probe_result, _run_result, _cli_module) do
    %{
      "response" => "",
      "model" => model || "cli-default",
      "status" => "not_found",
      "exit_code" => nil,
      "exit_signal" => nil,
      "stderr" => "#{cli_name} CLI not found in PATH",
      "elapsed_ms" => 0,
      "parse_error" => nil,
      "truncated" => false,
      "stderr_truncated" => false,
      "session_id" => nil
    }
  end

  def build_result(cli_name, _path, model, %{alive: false} = probe_result, nil, _cli_module) do
    reason = Map.get(probe_result, :reason, "unknown")
    cli_lower = String.downcase(cli_name)

    %{
      "response" => "",
      "model" => model || "cli-default",
      "status" => "probe_failed",
      "exit_code" => Map.get(probe_result, :exit_code),
      "exit_signal" => nil,
      "stderr" =>
        "#{cli_name} CLI probe failed: #{reason}. Run #{cli_lower} --version to diagnose.",
      "elapsed_ms" => 0,
      "parse_error" => nil,
      "truncated" => false,
      "stderr_truncated" => false,
      "session_id" => nil
    }
  end

  def build_result(_cli_name, _path, model, _probe_result, raw, cli_module) do
    parsed = cli_module.parse_output(raw.stdout, raw.stderr)

    timeout_message =
      if raw.timed_out do
        timeout_seconds = max(1, div(raw.elapsed_ms + 999, 1_000))

        "Request timed out after #{timeout_seconds}s. Retry with a longer timeout or resume the session."
      end

    status =
      cond do
        raw.timed_out -> "timeout"
        raw.exit_signal != nil -> "terminated"
        raw.exit_code != nil and raw.exit_code != 0 and parsed.status == :ok -> "error"
        true -> to_string(parsed.status)
      end

    model_used = Map.get(parsed.metadata, :model_used) || model || "cli-default"

    %{
      "response" => timeout_message || parsed.response,
      "model" => model_used,
      "status" => status,
      "exit_code" => raw.exit_code,
      "exit_signal" => raw.exit_signal,
      "stderr" => raw.stderr,
      "elapsed_ms" => raw.elapsed_ms,
      "parse_error" => if(raw.timed_out, do: nil, else: parsed.parse_error),
      "truncated" => raw.truncated,
      "stderr_truncated" => raw.stderr_truncated,
      "session_id" => parsed.session_id
    }
  end

  @spec build_meta(results :: map(), cli_configs :: [map()]) :: map()
  def build_meta(results, cli_configs) do
    files =
      case List.first(cli_configs) do
        %{files: files} -> files
        _ -> []
      end

    roles =
      Enum.reduce(cli_configs, %{}, fn cfg, acc ->
        Map.put(acc, "#{cfg.name}_role", cfg.role)
      end)

    total_elapsed_ms =
      results
      |> Map.delete("meta")
      |> Map.values()
      |> Enum.map(&(Map.get(&1, "elapsed_ms") || 0))
      |> Enum.max(fn -> 0 end)

    %{
      "total_elapsed_ms" => total_elapsed_ms,
      "files_referenced" => files
    }
    |> Map.merge(roles)
  end

  @spec encode(map()) :: String.t()
  def encode(results) do
    Jason.encode!(results, pretty: true)
  end
end

File: internal/roundtable/output.go
package roundtable

import "fmt"

// RawRunOutput holds the raw output from a subprocess execution,
// before any CLI-specific parsing. Populated by the subprocess runner
// (to be implemented in Phase 2C).
type RawRunOutput struct {
	Stdout          []byte  // raw stdout bytes
	Stderr          string  // stderr (may be truncated)
	ExitCode        *int    // nil if process was signaled
	ExitSignal      *string // e.g. "SIGKILL", nil if exited normally
	TimedOut        bool    // true if the process was killed due to timeout
	ElapsedMs       int64   // wall-clock milliseconds
	Truncated       bool    // true if stdout was capped
	StderrTruncated bool    // true if stderr was capped
}

// ParsedOutput holds the result of CLI-specific output parsing.
// Each backend parser (Gemini, Claude, Codex) produces this struct
// from the raw stdout/stderr. Parsers will be implemented in Phase 2C.
type ParsedOutput struct {
	Response   string         // extracted response text
	Status     string         // "ok", "error", "rate_limited", etc.
	ParseError *string        // non-nil if parsing failed
	SessionID  *string        // session ID extracted from output (Claude)
	Metadata   map[string]any // includes "model_used", "tokens", "usage", etc.
}

// BuildResult normalizes raw subprocess output and parsed CLI output into
// a Result that matches the Elixir output.ex JSON contract.
//
// This implements the third clause of build_result/6 — the normal run path.
// The not_found and probe_failed cases are handled by NotFoundResult and
// ProbeFailedResult in result.go.
//
// Status priority:
//  1. TimedOut   → "timeout"  (overrides everything)
//  2. ExitSignal → "terminated"
//  3. Non-zero exit + parsed status "ok" → "error"
//  4. Otherwise  → parsed status as-is
func BuildResult(raw RawRunOutput, parsed ParsedOutput, fallbackModel string) *Result {
	// Compute status
	status := parsed.Status
	switch {
	case raw.TimedOut:
		status = "timeout"
	case raw.ExitSignal != nil:
		status = "terminated"
	case raw.ExitCode != nil && *raw.ExitCode != 0 && parsed.Status == "ok":
		status = "error"
	}

	// Compute response — timeout message overrides parsed response
	response := parsed.Response
	if raw.TimedOut {
		timeoutSeconds := max((raw.ElapsedMs+999)/1000, 1)
		response = fmt.Sprintf(
			"Request timed out after %ds. Retry with a longer timeout or resume the session.",
			timeoutSeconds,
		)
	}

	// Compute model — prefer parsed Metadata, fall back to request-level, then default
	var model string
	if m, ok := parsed.Metadata["model_used"].(string); ok && m != "" {
		model = m
	}
	if model == "" {
		model = fallbackModel
	}
	if model == "" {
		model = "cli-default"
	}

	// Clear parse_error on timeout — the response is synthetic, not parsed
	var parseError *string
	if !raw.TimedOut {
		parseError = parsed.ParseError
	}

	return &Result{
		Response:        response,
		Model:           model,
		Status:          status,
		ExitCode:        raw.ExitCode,
		ExitSignal:      raw.ExitSignal,
		Stderr:          raw.Stderr,
		ElapsedMs:       raw.ElapsedMs,
		ParseError:      parseError,
		Truncated:       raw.Truncated,
		StderrTruncated: raw.StderrTruncated,
		SessionID:       parsed.SessionID,
	}
}

// BuildMeta computes aggregate metadata across all results in a roundtable
// session. Ports build_meta/2 from output.ex.
//
// TotalElapsedMs is the max of all results' ElapsedMs (wall-clock of the
// slowest participant). FilesReferenced and DynamicFields (role names) are
// passed through as-is.
func BuildMeta(results map[string]*Result, files []string, roles map[string]string) Meta {
	var maxElapsed int64
	for _, r := range results {
		if r.ElapsedMs > maxElapsed {
			maxElapsed = r.ElapsedMs
		}
	}
	if files == nil {
		files = []string{}
	}
	return Meta{
		TotalElapsedMs:  maxElapsed,
		FilesReferenced: files,
		DynamicFields:   roles,
	}
}
